src/kaleidoscope/device/dygma/Raise.cpp

```diff
@@ -0,0 +1,599 @@
+  for (byte row = 0; row < 5; row++) {
+    for (byte col = 0; col < 8; col++) {
+      uint8_t keynum = (row * 8) + (col);
+      uint8_t keyState;
+
+      // left
+      keyState = (bitRead(previousLeftHandState.all, keynum) << 0) |
+                 (bitRead(leftHandState.all, keynum) << 1);
+      if (keyState)
+        ThisType::handleKeyswitchEvent(Key_NoKey, KeyAddr(row, col), keyState);
+
+      // right
+      keyState = (bitRead(previousRightHandState.all, keynum) << 0) |
+                 (bitRead(rightHandState.all, keynum) << 1);
+      if (keyState)
+        ThisType::handleKeyswitchEvent(Key_NoKey, KeyAddr(row, 15 - col), keyState);
+    }
+  }
+}
+
+void RaiseKeyScanner::scanMatrix() {
+  readMatrix();
+  actOnMatrixScan();
+}
+
+void RaiseKeyScanner::maskKey(KeyAddr key_addr) {
+  if (!key_addr.isValid())
+    return;
+
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    rightHandMask.rows[row] |= 1 << (7 - (col - 8));
+  } else {
+    leftHandMask.rows[row] |= 1 << (7 - col);
+  }
+}
+
+void RaiseKeyScanner::unMaskKey(KeyAddr key_addr) {
+  if (!key_addr.isValid())
+    return;
+
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    rightHandMask.rows[row] &= ~(1 << (7 - (col - 8)));
+  } else {
+    leftHandMask.rows[row] &= ~(1 << (7 - col));
+  }
+}
+
+bool RaiseKeyScanner::isKeyMasked(KeyAddr key_addr) {
+  if (!key_addr.isValid())
+    return false;
+
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    return rightHandMask.rows[row] & (1 << (7 - (col - 8)));
+  } else {
+    return leftHandMask.rows[row] & (1 << (7 - col));
+  }
+}
+
+void RaiseKeyScanner::maskHeldKeys() {
+  memcpy(leftHandMask.rows, leftHandState.rows, sizeof(leftHandMask));
+  memcpy(rightHandMask.rows, rightHandState.rows, sizeof(rightHandMask));
+}
+
+bool RaiseKeyScanner::isKeyswitchPressed(KeyAddr key_addr) {
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    return (bitRead(rightHandState.rows[row], 15 - col) != 0);
+  } else {
+    return (bitRead(leftHandState.rows[row], col) != 0);
+  }
+}
+
+bool RaiseKeyScanner::wasKeyswitchPressed(KeyAddr key_addr) {
+  auto row = key_addr.row();
+  auto col = key_addr.col();
+
+  if (col >= 8) {
+    return (bitRead(previousRightHandState.rows[row], 15 - col) != 0);
+  } else {
+    return (bitRead(previousLeftHandState.rows[row], col) != 0);
+  }
+}
+
+uint8_t RaiseKeyScanner::pressedKeyswitchCount() {
+  return __builtin_popcountl(leftHandState.all) + __builtin_popcountl(rightHandState.all);
+}
+
+uint8_t RaiseKeyScanner::previousPressedKeyswitchCount() {
+  return __builtin_popcountl(previousLeftHandState.all) + __builtin_popcountl(previousRightHandState.all);
+}
+
+void RaiseKeyScanner::setKeyscanInterval(uint8_t interval) {
+  RaiseHands::leftHand.setKeyscanInterval(interval);
+  RaiseHands::rightHand.setKeyscanInterval(interval);
+}
+
+void RaiseKeyScanner::setup() {
```

Comment:
Lots of magic numbers.

Reply:
Not quite sure what I can do about the magic numbers there... hmm... I can have something like `static constexpr uint8_t keyscanner_pins[] = {2, 3, ....}`, and iterate over that instead of two for loops with magic numbers.